lib/maths/CBoostedTreeImpl.cc

```diff
@@ -354,23 +359,35 @@ CBoostedTreeImpl::candidateSplits(const core::CDataFrame& frame,
     }
     for (std::size_t i = 0; i < features.size(); ++i) {
 
-        TDoubleVec columnSplits;
-        columnSplits.reserve(m_NumberSplitsPerFeature - 1);
+        TDoubleVec featureSplits;
+        featureSplits.reserve(m_NumberSplitsPerFeature - 1);
 
         for (std::size_t j = 1; j < m_NumberSplitsPerFeature; ++j) {
             double rank{100.0 * static_cast<double>(j) /
                         static_cast<double>(m_NumberSplitsPerFeature)};
             double q;
-            if (columnQuantiles[i].quantile(rank, q)) {
-                columnSplits.push_back(q);
+            if (featureQuantiles[i].quantile(rank, q)) {
+                featureSplits.push_back(q);
             } else {
                 LOG_WARN(<< "Failed to compute quantile " << rank << ": ignoring split");
             }
         }
 
-        columnSplits.erase(std::unique(columnSplits.begin(), columnSplits.end()),
-                           columnSplits.end());
-        candidateSplits[features[i]] = std::move(columnSplits);
+        const auto& dataType = m_FeatureDataTypes[features[i]];
+
+        if (dataType.s_IsInteger) {
+            std::for_each(featureSplits.begin(), featureSplits.end(),
+                          [](double& split) { split = std::floor(split) + 0.5; });
```

Comment:
I am not sure why we are doing a floor and then adding `0.5`. Why do we always want to split halfway on an integer? It seems to me that this is the exact same thing as calling `ceil` against the `split`, especially if all the data are integers.

Reply:
The rationale is mainly to snap splits to the same values so we don't needless "duplicate". When we compute quantiles we linearly interpolate so we tend to end up with many distinct values which result in identical partitions of the data set. By snapping to a value (the midpoint is a good as any other) when we subsequently remove duplicates we remove all splits which we know must be equivalent because they fall between two consecutive integers. (I could be cleverer here and actually check to see if there are values between candidate splits, but this requires an additional pass over the data so the runtime tradeoff becomes less clear especially since it depends on the feature distribution.)

Reply:
Do you think it'd be helpful to add a comment @benwtrent?

Reply:
Yeah, I see that, a `1.1` is the same as a `1.2` as both `1.0` is less than and `2.0` is greater than. Having a comment explaining the "magical number" would be nice. 

Something like
// Since we are splitting Integers, all decimal values incur the same branch path. This allows  duplicate splits to be pruned

Reply:
See [fa42136](https://github.com/elastic/ml-cpp/pull/609/commits/fa421361eca9bf84d83d7cdeaa6859c547ada88a).